### PR TITLE
feat: Prefix Delegation

### LINKED ISSUE #5478: Add Windows support for prefix delegation
### Description

**What problem are you trying to solve?**
This feature aims to improve pod density capabilities for Windows EKS nodes launched with Karpenter. Currently, by default, Karpenter launches Windows pods with a setting of `--max-pods=110` meaning a node cannot scale beyond 110 pods. 

**How important is this feature to you?**
This is an important capability because using prefix delegation will offer improved cost savings, since more pods can be scheduled on a node.

**References**
- [Increasing pod density for Windows nodes on Amazon EKS](https://aws.amazon.com/blogs/containers/increasing-pod-density-for-windows-nodes-on-amazon-eks/)

* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by CorentinPtrl**
feat: Prefix Delegation

Fixes #5478

**Description**
Support prefix delegation allowing proper max-pods calculations

**How was this change tested?**

**Does this change impact docs?**
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by archer-65** [comment]
Any issues with this feature? We currently set maxPods to 110 as workaround, but we'd prefer to let Karpenter calculate the correct number for us.

**Comment by CorentinPtrl** [comment]
No there is no issues with this feature

**Comment by archer-65** [comment]
@tzneal can this be reopened then?

File: pkg/providers/instancetype/types.go [hunk 1]
```diff
@@ -336,19 +336,39 @@ func efas(info *ec2.InstanceTypeInfo) *resource.Quantity {
...
+			return resource.NewQuantity(0, resource.DecimalSI)
+		}
+		addressesPerInterface := *info.NetworkInfo.Ipv4AddressesPerInterface
+		count := (usableNetworkInterfaces * (addressesPerInterface - 1) * 16) + 2
+		//Limit the total number of pods that can be launched on any instance type based on the vCPUs on that instance type.
+		if *info.VCpuInfo.DefaultVCpus > 30 {
+			return resources.Quantity(fmt.Sprint(min(250, count)))
+		} else {
+			return resources.Quantity(fmt.Sprint(min(110, count)))
+		}
```
Comment (grosser) [inline pkg/providers/instancetype/types.go:355]: FYI
```
	count = min(count, int64(lo.Ternary(lo.FromPtr(info.VCpuInfo.DefaultVCpus) > 30, 250, 110)))
	return resources.Quantity(fmt.Sprint(count))
```

**Comment by grosser** [comment]
FYI https://github.com/aws/karpenter-provider-aws/pull/8406 is a stripped down + dumbed down + compiling version of this patch that should be easier to maintain on a fork

**Comment by CorentinPtrl** [comment]
@grosser 
Your pull request doesn’t seem to account for the possibility that prefix delegation might be disabled, nor does it verify whether the instance type actually supports prefix delegation.

**Comment by grosser** [comment]
thx, yeah that's on purpose so we our fork is simpler, we always use it + only have instances that support it (don't have super or old or bare metal ones)

**Comment by CorentinPtrl** [comment]
Maybe we can work together on this PR if needed I’m not sure your version should be merged as-is, since it could break clusters for people without prefix delegation or on unsupported instance types

**Comment by grosser** [comment]
afaik this PR does not need work ... just needs to be merged 😭

**Comment by CorentinPtrl** [comment]
@saurav-agarwalla, since you’re assigned to #8210, could you please review it or approve it?

**Comment by velkovb** [comment]
We are looking for the same functionality, any chance this can be revived?

**Pull request closed without merging**